Task 3, 4 requirements

din-carmon opened the pull request:
- Visitor interface support for fun programs
- Simple example visitors: int visitor, error detectors, evaluation visitor
- Relevant tests

din-carmon committed:
Task 3, 4 requirements:
- Visitor interface support for fun programs
- Simple example visitors: int visitor, error detectors, evaluation visitor
- Relevant tests

ronhass-classiq commented on src/fun_language_package/fun_visitor.py:
```diff
@@ -0,0 +1,27 @@
+from typing import Any
+
+
+class FunVisitor:
```
Very nice!

Another visitor implementation you will see in our codebase doesn't use the `accept` methods, but implements visit on a node by calling the function `f"visit_{type(node).__name__}"`, so for example here you would have `visit_FunInt`, `visit_FunVar` etc.

No need to change anything, just wanted to let you know about this approach.

din-carmon replied:
Nice :)

ronhass-classiq commented on src/fun_language_package/fun_visitor.py:
```diff
@@ -0,0 +1,27 @@
+from __future__ import annotations  # 3.11+ does this by default
```
You can assume python 3.11+ if you want (and update this assumption in `pyproject.toml`). But you can leave it as it is.

din-carmon replied:
Fixed

din-carmon replied:
Apparently even in python 3.11 it is needed for the runtime interpreter. reverting

ronhass-classiq commented on src/fun_language_package/unbound_variable_detector_visitor.py:
```diff
@@ -0,0 +1,37 @@
+        bounded_variables_copy = self.bounded_variables.copy()
+        self.bounded_variables.add(fun_function_call.function_expression.fun_var.var_name)
+        fun_function_call.function_expression.accept(self)
+        self.bounded_variables = bounded_variables_copy
```
It's not very important in this case, but it's a good paradigm to use `try: ... finally: ...` when you make a change to a context and then want to revert it always.

din-carmon replied:
Fixed

ronhass-classiq commented on src/fun_language_package/function_argument_is_function_detector_visitor.py:
```diff
@@ -0,0 +1,12 @@
+    def visit_fun_function_call(self, fun_function_call: fun_utils.FunFunctionCall) -> None:
+        fun_function_call.function_expression.accept(self)
+
+        if isinstance(fun_function_call.fun_argument, fun_utils.FunFunction):
```
You can do the validation first and then just call `super`

din-carmon replied:
fixed

ronhass-classiq commented on src/fun_language_package/evaluation_visitor.py:
```diff
@@ -0,0 +1,69 @@
+        tmp_copy_of_current_assignments = self.assignments.copy()
+
+        left = fun_plus_operation.left.accept(self)
+        self.assignments = tmp_copy_of_current_assignments.copy()  # Reset assignments
```
I don't think this is necessary. You only change `assignments` in a function call, and you revert it right after the evaluation of the call. 
After visiting a node you will always get back to the original assignments, as every `visit` function preserves them.

din-carmon replied:
fixed

ronhass-classiq commented on src/fun_language_package/evaluation_visitor.py:
```diff
@@ -0,0 +1,69 @@
+
+        # 1. Evaluate the argument.
+        argument = fun_function_call.fun_argument.accept(self)
+        self.assignments = tmp_copy_of_current_assignments.copy() # Reset assignments
```
See last comment (I think it's unnecessary)

din-carmon replied:
fixed

ronhass-classiq commented on src/fun_language_package/evaluation_visitor.py:
```diff
@@ -0,0 +1,69 @@
+        ret = fun_function_call.function_expression.fun_expression.accept(self)
+        self.assignments = tmp_copy_of_current_assignments.copy() # Reset assignments
+
+        if not isinstance(ret, fun_utils.FunInt):
```
You already check this in a different visitor, I think (do you have a test case when the other visitor passes and this validation fails?)

ronhass-classiq replied:
`FunctionReturnFunctionDetectorVisitor` doesn't catch this?

din-carmon replied:
Oh I see what you meant previously now. But is it a valid assumption that the user ran other visitors on the program before evaluation?

din-carmon replied:
ok. Now i saw your comment on invocation of the other visitors in the evaluation function.
Makes sense. fixing

ronhass-classiq commented on tests/test_evaluation_visitor.py:
```diff
@@ -0,0 +1,73 @@
+from fun_language_package import evaluation_visitor
```
I think you should already change your `evaluate` function to use `EvaluationVisitor` (after all the validations visitors) and remove the old `evaluate` code. Then, you won't need this test file, as it's already covered in the existing tests.

din-carmon replied:
fixed

ronhass-classiq reviewed (commented):
Really nice, I don't see any issues, just some minor comments.

din-carmon committed:
PR fixes:
- Removed the previous evaluation mechanism. Now it is done only using the evaluation visitor
- Some other minor fixes

ronhass-classiq commented on src/fun_language_package/evaluation_visitor.py:
```diff
@@ -0,0 +1,60 @@
+        self.assignments: dict[str, fun_utils.FunProgram] = {}
+
+    def visit(self, program: tuple[Any, ...] | int) -> int:
+        fun_program = fun_utils.build_fun_program(program)
```
Use `super().visit` here

ronhass-classiq commented on src/fun_language_package/fun_utils.py:
```diff
@@ -162,9 +138,5 @@ def build_fun_program(program: tuple[Any, ...] | int | str) -> FunProgram:
-        return evaluation.value
-
-    raise EvaluationIsNotIntError(evaluation)
+    visitor = evaluation_visitor.EvaluationVisitor()
```
Call the validation visitors before

din-carmon replied:
fixed

ronhass-classiq commented on pyproject.toml:
```diff
@@ -38,7 +38,11 @@ strict = true
 [tool.ruff]
 lint.select = ["F", "E", "W", "UP", "B", "C90", "N", "PL", "SIM"]  # include pyflakes rules
-lint.ignore = ["PLR2004"] # Ignore magic values in code
+lint.ignore = [
```
General note about ignores: you should think whether the ignore is something general/global to your project, or actually just fits one line. In the latter case, it's best to ignore this specific line. This is how we do it in Cadmium.

din-carmon replied:
Got it :)
Would you like me to fix that? Or is it just a note?

ronhass-classiq replied:
That's OK. let's call it a day :)

din-carmon committed:
PR fixes.
- Major change: In evaluation function - run the error detectors visitors first
